1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  

```

Comment:
I think it complicates the ForEach-Object cmdlet too much and limits future development opportunities.

From previous discussion https://github.com/PowerShell/PowerShell-RFC/pull/174#issuecomment-492906471
> From iSazonov 

Yesterday I accidentally saw GNU utility like "parallel --pipe". Also we could find "Where-Object -Parallel" useful too. And "Select-String". And others. So this suggests that maybe we need something more general for pipe parallelization (Parallel-Pipe)
$hugeArray | Parallel-Pipe { Where-Object Name -eq "test" }
$hugeArray | Parallel-Pipe {Select-String -Pattern "test" }
$hugeArray | Parallel-Pipe -AsJob { ... }

> From rjmholt 

A rose by any other name would smell as sweet:
$hugeArray | ForEach-Object -Parallel { $_ | Where-Object Name -eq "test" }
$hugeArray | ForEach-Object -Parallel { $_ | Select-String -Pattern "test" }
$hugeArray | ForEach-Object -Parallel -AsJob { ... }
I think in the name of single-purpose, composible cmdlets having a single parallel cmdlet (rather than trying to parallelise each individually) is the right way to go. But I think ForEach-Object has that brief of Select-like enumeration. And just like LINQ has .AsParallel(), I think -Parallel makes sense to do this. But that's just a brief and not-strongly-held opinion :)

> From iSazonov

There is a problem in depth with "ForEach-Object -Parallel" - there is a lot of parameters and how will "-Begin/Process/End" and etc work - in "ForEach-Object -Parallel" context or scriptblock context? I think we will have to duplicate parameters. In the case Parallel-Pipe is more simple and more safely.
Parallel-Pipe     -InputOblect $data
                  -Begin { … } 
                  -Process { Foreach-Object -Begin { … } -Process { … } -End { … } }
                  -End { … }

Reply:
-Begin, -Process, -End switches will not be part of the new parameter set, since they don't make sense in this parallel case.

Reply:
There's a meaning. Especially in binary cmdlets It should work just like in ForEach-Object (why do we have the blocks there?). Rather, there is a difficulty in implementation. 
Current design looks like a workaround - it is simple implement but save all current limitations.. My suggestion is to implement natively parallel pipelines that give significant flexibility, performance and extensibility.

Reply:
I think it does make sense to keep the 
```
before any {do this }
{do this for each one}
after all of them  {do this} 
```
that exists in the sequential for each.

Reply:
If we're removing `-Begin`, `-Process`, and `-End`  then why are you proposing to overload `ForEach-Object` instead of simply adding a new command? There's _literally_ nothing else in this command, so you're really creating an entirely new command!

`ForEach-Object` is currently one of the worst performing cmdlets in PowerShell (compare `$power = 1..1e5 | foreach { $_ * 2 }` to `$power = 1..1e5 | & { process { $_ * 2 } }`) and although I doubt this change will make it slower, it _will_ make it more complicated -- reducing the possibility of ever fixing it 🤨